Auth signin and signup with db

Abas-dev opened the pull request:
This has the db tables add and the user registration and login worked on.

bolexs commented:
@Abas-dev kindly follow the pr template convention given

bolexs commented on api/v1/routes/auth.py:
i can't see any tests, where are your tests @Abas-dev??

Abas-dev replied:
@bolexs 

pls check 

I created a tempoary register and login endpoint for those who depend on it to start work and updated the required tables the devs want

Abas-dev commented:
<!--- Provide a general summary of your changes in the Title above -->

​

## Added temp register and login and updated the db tables

In this repo, i added a register and login endpoint to allow users create and login an account. 

I also updated all tables which have been created by the db team so as to allow others work on their endpoint faster.

​

## Related Issue ([Link to issue ticket](https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/20))

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

​

## Motivation and Context

The change is required to allow other devs access to register, login and db tables

​

## How Has This Been Tested?

Its been tested using postman 

​

## Screenshots (if appropriate - Postman, etc):
![Screenshot (161)](https://github.com/user-attachments/assets/add404f8-3410-4408-ac48-042bc51b484e)

​

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [x ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
      ​

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ x] All new and existing tests passed.

bolexs commented on api/v1/routes/auth.py:
```diff
@@ -0,0 +1,105 @@
+    db.commit()
+    db.refresh(db_user)
+    return db_user
+    # user_response = UserResponse.from_orm(db_user)
```
you need to clean up, no comments allowed. If the code is not in use it shouldn't be there

Abas-dev commented:
@bolexs 

done that

bolexs reviewed (commented):
i need you clean up the codebase and explain in ur pr how i can test everything you have done

**Also add tests for your said implementations**

bolexs commented:
> ​
> 
> ## Added temp register and login and updated the db tables
> In this repo, i added a register and login endpoint to allow users create and login an account.
> 
> I also updated all tables which have been created by the db team so as to allow others work on their endpoint faster.
> 
> ​
> 
> ## Related Issue ([Link to issue ticket](https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/20))
> ​
> 
> ## Motivation and Context
> The change is required to allow other devs access to register, login and db tables
> 
> ​
> 
> ## How Has This Been Tested?
> Its been tested using postman
> 
> ​
> 
> ## Screenshots (if appropriate - Postman, etc):
> ![Screenshot (161)](https://private-user-images.githubusercontent.com/79318104/350706987-add404f8-3410-4408-ac48-042bc51b484e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE0OTQyNzEsIm5iZiI6MTcyMTQ5Mzk3MSwicGF0aCI6Ii83OTMxODEwNC8zNTA3MDY5ODctYWRkNDA0ZjgtMzQxMC00NDA4LWFjNDgtMDQyYmM1MWI0ODRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIwVDE2NDYxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0M2ZlZTk0ZGJmNjJkNjM4MzhkMjJiMzYyYTNiNzg4MmNhY2RkYzI1ZTM3MjRlOGI2NDI2Yjk1MmMyYzA1MWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Rdi-uBWN3_tau34FY5kszG6X1540r5X1GwGIHmqW5EE)
> 
> ​
> 
> ## Types of changes
> * [x ] Bug fix (non-breaking change which fixes an issue)
> * [x ] New feature (non-breaking change which adds functionality)
> * [x]  Breaking change (fix or feature that would cause existing functionality to change)
>   ​
> 
> ## Checklist:
> * [ ]  My code follows the code style of this project.
> * [x]  My change requires a change to the documentation.
> * [ ]  I have updated the documentation accordingly.
> * [x ] I have read the **CONTRIBUTING** document.
> * [ ]  I have added tests to cover my changes.
> * [ x] All new and existing tests passed.

add link to ur issues and add a response to the other sections

Abas-dev commented:
@bolexs 

this is not the final ommit 

i am going to work on the endpoint 

its just temp 

it works 

i have shown you the screanshot

bolexs commented:
> @bolexs
> 
> this is not the final ommit
> 
> i am going to work on the endpoint
> 
> its just temp
> 
> it works
> 
> i have shown you the screanshot

what do you mean by temp?

Abas-dev commented:
@bolexs 
i have a low pc 

pls this is for the db and for the devs to have the db to work on

Abas-dev commented:
Temporary 

i will still work on it again 

to re-arrage it properly.

Abas-dev commented:
@bolexs 

pls respond

bolexs commented:
> @bolexs
> 
> pls respond

can you attach the link to ur issue

Abas-dev commented:
i have done that

Abas-dev commented:
check the PR temp  i wrote

Dev-wonderful commented on api/utils/auth.py:
```diff
@@ -0,0 +1,77 @@
+from api.db.database import get_db
+
+# Define your JWT secret and algorithm
+SECRET_KEY = os.getenv("SECRET_KEY", "MY SECRET KEY")
+ALGORITHM = os.getenv("ALGORITHM", "HS256")
```
It's better to put all env variables in a single file, and import from there. There's a way to do that in FastApi, but you can just move all envs to a single file for now

Dev-wonderful commented on api/utils/auth.py:
```diff
@@ -0,0 +1,77 @@
+    encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
+    return encoded_jwt
+
+def get_current_user(db: Session = Depends(get_db), token: str = Depends(oauth2_scheme)):
+    credentials_exception = HTTPException(
+        status_code=status.HTTP_401_UNAUTHORIZED,
+        detail="Could not validate credentials",
+        headers={"WWW-Authenticate": "Bearer"},
+    )
+    try:
+        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
+        username: str = payload.get("username")
+        if username is None:
+            raise credentials_exception
+        token_data = TokenData(username=username)
+    except JWTError:
+        raise credentials_exception
+    user = db.query(User).filter(User.username == token_data.username).first()
+    if user is None:
+        raise credentials_exception
+    return user
+
+def get_current_admin(db: Session = Depends(get_db), token: str = Depends(oauth2_scheme)):
+    user = get_current_user(db, token)
+    if not user.is_admin:
+        raise HTTPException(
+            status_code=status.HTTP_403_FORBIDDEN,
+            detail="You do not have permission to access this resource",
+        )
+    return user
```
This should be in a dependencies file

Dev-wonderful commented on api/v1/models/invitation.py:
```diff
@@ -0,0 +1,18 @@
+from sqlalchemy.dialects.postgresql import UUID
+
+# Invitation model
+class Invitation(Base):
+    __tablename__ = 'invitations'
+
+    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid7)
+    user_id = Column(UUID(as_uuid=True), ForeignKey('users.id', ondelete='CASCADE'), nullable=False)
+    organization_id = Column(UUID(as_uuid=True), ForeignKey('organizations.id', ondelete='CASCADE'), nullable=False)
+    created_at = Column(DateTime(timezone=True), server_default=func.now())
+    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
+    expires_at = Column(DateTime(timezone=True), nullable=False)
+    is_valid = Column(Boolean, default=True)
```
An abstract BaseModel class was created to give every model a set of base fields and functionality, inherit from that and remove duplicate fields --- Do this for all classes

Abas-dev replied:
when merged 

sunday will fix it 

i dont know anything here

Dev-wonderful commented on api/v1/routes/auth.py:
```diff
@@ -0,0 +1,89 @@
+db = next(get_db())
+
+
+router = APIRouter()
+
+ACCESS_TOKEN_EXPIRE_MINUTES = 30
+
+@router.post("/login", response_model=Token,status_code=status.HTTP_200_OK)
```
Give the router a proper semantic name, and add a `/auth` prefix and `auth` tag for swagger docs. to the `APIRouter` class

Also create a base router for `/api/v1`, and include other routers to it

Abas-dev replied:
i dont know what you mean here

Dev-wonderful commented on api/v1/routes/auth.py:
```diff
@@ -0,0 +1,89 @@
+    )
+    return {"access_token": access_token, "token_type": "bearer"}
+  
+@router.post("/register", response_model=UserCreate, status_code=status.HTTP_201_CREATED)
+def register_user(user: UserCreate, db: Session = Depends(get_db)):
```
create a `scoped_session` for db, and import wherever it is needed, this method would cause unintended consequences if the endpoint is called with a `db` query parameter

Dev-wonderful commented on api/v1/routes/auth.py:
```diff
@@ -0,0 +1,89 @@
+    return db_user
+    
+
+# Protected route example: test route
+@router.get("/admin")
+def read_admin_data(current_admin: User = Depends(get_current_admin)):
```
Use `Annotated[User, Depends(get_current_admin)]` for this, that's how to properly use dependencies.
Update this all round

Abas-dev replied:
> Use `Annotated[User, Depends(get_current_admin)]` for this, that's how to properly use dependencies. Update this all round

fixed this

Dev-wonderful commented on api/v1/schemas/auth.py:
```diff
@@ -0,0 +1,23 @@
+import uuid
+
+
+class UserCreate(BaseModel):
+    username: str
+    email: str
+    password: str
+    first_name: str | None = None
+    last_name: str | None = None
+
```
`passowrd` is not being properly validated

Abas-dev replied:
> `passowrd` is not being properly validated

i have fixed this

Dev-wonderful commented on main.py:
```diff
@@ -32,7 +33,7 @@ async def lifespan(app: FastAPI):
 )
 
 
-# app.include_router(auth, tags=["Auth"])
+app.include_router(auth, tags=["Auth"],prefix="/api/auth/v1")
 # app.include_router(users, tags=["Users"])
```
Fix this based on the previous comment about the router

Abas-dev commented:
@Dev-wonderful 

i know 

but this is something temp 

if this is merged 

sunday and the rest can fix the issue

this is so that others can work

and i have a low battery
